business/mesh.go

```diff
@@ -0,0 +1,315 @@
+package business
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/log"
+)
+
+// MeshService is a support service for retrieving data about the mesh environment
+// when Istio is installed with multi-cluster enabled. Prefer initializing this
+// type via the NewMeshService function.
+type MeshService struct {
+	k8s kubernetes.ClientInterface
+
+	// newRemoteClient is a helper variable holding a function that should return an
+	// initialized kubernetes client using the specified config argument. This was created,
+	// mainly, for tests to set a function returning a mock of the kuberentes client.
+	newRemoteClient func(config *rest.Config) (kubernetes.ClientInterface, error)
+}
+
+// Cluster holds some metadata about a cluster that is
+// part of the mesh.
+type Cluster struct {
+	// ApiEndpoint is the URL where the Kubernetes/Cluster API Server can be contacted
+	ApiEndpoint string `json:"apiEndpoint"`
+
+	// IsKialiHome specifies if this cluster is hosting this Kiali instance (and the observed Mesh Control Plane)
+	IsKialiHome bool `json:"isKialiHome"`
+
+	// Name specifies the CLUSTER_ID as known by the Control Plane
+	Name string `json:"name"`
+
+	// Network specifies the logical NETWORK_ID as known by the Control Plane
+	Network string `json:"network"`
+
+	// SecretName is the name of the kubernetes "remote secret" where data of this cluster was resolved
+	SecretName string `json:"secretName"`
+}
+
+// NewMeshService initializes a new MeshService structure with the given k8s client and
+// newRemoteClientFunc arguments (see the MeshService struct for details). The newRemoteClientFunc
+// can be passed a nil value and a default function will be used.
+func NewMeshService(k8s kubernetes.ClientInterface, newRemoteClientFunc func(config *rest.Config) (kubernetes.ClientInterface, error)) MeshService {
+	if newRemoteClientFunc == nil {
+		// Mesh discovery will use ServiceAccount token instead of user token
+		token := ""
+		token, _ = kubernetes.GetKialiToken()
+
+		newRemoteClientFunc = func(config *rest.Config) (kubernetes.ClientInterface, error) {
+			return kubernetes.NewClientFromConfig(config, &api.AuthInfo{Token: token})
```

Comment:
This is misleading. We have to honor the token that is passed in `config` argument. But I tried the code and it's working anyway.

I reviewed what's inside that `NewClientFromConfig` and, indeed, it does nothing with the token that is passed to it via `authInfo` other than simply storing for later. The real client is created purely from `config`. 

I have the idea that current state of `NewClientFromConfig` is error prone, because it ignores the token, but it can merge Impersonation data regardless of any token mismatch between `config` and `authInfo`.

:/

Reply:
For more context: this piece of code is creating a kubernetes client for remote clusters; not the "home/local" cluster. So, any local token or auth context that there is in place is not valid for querying the remote clusters API Servers.

Reply:
@israel-hdez I agree with your comments.  This initial support was only an attempt to fix the rebase issues for compilation.  I'll resolve this conversation and you can push any update you feel is more suitable.

Reply:
@jshaughn I chose to create a PR to let @jmazzitelli and @mlbiam take a look.
Here it is: https://github.com/kiali/kiali/pull/3602